Refactor nactionscu

dwikler opened the pull request:
This merge request addresses part of issue https://github.com/sjswerdloff/tdwii_plus_examples/issues/85.
It refactors nactionscu.py from the tdwii_plus_examples folder.
The NActionSCU class is refactored as UPSPullNActionSCU in tdwii_plus_examples package folder in upspullnactionscu.py.
UPSPullNActionSCU is derived from a BaseSCU class.
The nactionscu CLI application in tdwii_plus_examples package folder is also refactored as upschangestatescu.py CLI application and moved to the tdwii_plus_examples/cli folder.
Tests are provided.
Finally, the GUI application in tdwii_plus_examples/tdd folder is refactored to use UPSPullNActionSCU class.

## Summary by Sourcery

Refactor the UPS (Unified Procedure Step) related classes and CLI application to improve modularity, separation of concerns, and enhance the implementation of the UPS Pull N-Action Service Class User (SCU).

New Features:
- Implement UPSPullNActionSCU class for managing UPS state changes
- Create upschangestatescu CLI application for UPS state management

Enhancements:
- Refactor BaseSCU to improve association and context handling
- Improve error handling and logging in DICOM service classes
- Add more robust transaction UID management for UPS operations

Tests:
- Add comprehensive unit and integration tests for UPSPullNActionSCU
- Enhance existing test coverage for BaseSCU and related classes

Chores:
- Reorganize project structure
- Improve code documentation

dwikler commented on tdwii_plus_examples/tests/test_basescu_unit.py:
```diff
@@ -109,7 +109,7 @@ def test_associate_rejection(self):
+        assoc_result = base_scu._associate(verbose=True)
         self.assertEqual(assoc_result.status, "Error")
         self.assertEqual(str(assoc_result.description), "Called AE parameters not set")
         self.test_logger.info(
```
I understand I can improve coverage but I would pass for now to keep moving forward and hopefully finish the refactoring.

dwikler replied:
@sourcery-ai issue

dwikler commented on tdwii_plus_examples/tests/test_basescu_unit.py:
```diff
@@ -95,7 +95,7 @@ def test_associate_rejection(self):
-            assoc_result = self.base_scu._associate()
+            assoc_result = self.base_scu._associate(verbose=True)
             self.assertEqual(assoc_result.status, "Error")
         self.test_logger.info("Test _associate with rejection")
```
I understand I can improve coverage but I would pass for now to keep moving forward and hopefully finish the refactoring.

dwikler replied:
@sourcery-ai issue

dwikler commented on tdwii_plus_examples/tests/test_upspullnactionscu_integration.py:
```diff
@@ -0,0 +1,171 @@
+from tdwii_plus_examples.upspushncreatescu import UPSPushNCreateSCU
+
+
+class TestUPSPullNActionSCUIntegration(unittest.TestCase):
```
These are tested in unit test

dwikler commented on tdwii_plus_examples/tests/cli/test_upschangestatescu.py:
```diff
@@ -0,0 +1,67 @@
+import tdwii_plus_examples.cli.upschangestatescu as upschangestatescu
+
+
+class TestUPSChangeStateSCUCLI(unittest.TestCase):
```
I understand I can improve coverage but I would pass for now to keep moving forward and hopefully finish the refactoring.

dwikler replied:
@sourcery-ai issue

sjswerdloff commented:
Ask sourcery to create issues for each of its suggestions so we capture them and move forward

dwikler commented on tdwii_plus_examples/basescu.py:
```diff
@@ -103,73 +111,126 @@ def __init__(
+            return self._validate_sop_classes(result, required_sop_classes)
 
-    def _all_contexts_accepted(self, assoc: Association) -> AssociationResult:
+    def _check_accepted_sop_classes(self, assoc: Association) -> AssociationResult:
```
I want to keep both options to return simple boolean or detailed result information for now. We could create an improvement issue to provide this in all subclasses or to only keep boolean/int return values and propose another mechanism to retrieve the detailed result. The benefit of the detailed result vs log messages is mainly for GUI apps where meaningful messages could be displayed to user in case of errors.

dwikler replied:
@sourcery-ai issue

sjswerdloff commented on tdwii_plus_examples/upspullnactionscu.py:
```diff
@@ -0,0 +1,226 @@
+# Action Type ID constants
+ACTION_TYPE_ID_CHANGEUPSSTATE = 1
+
+# UPS Procedure Step State constants
```
could be an enum, but then you'd have to have IN_PROGRESS.value for the string.
This is fine, just thinking about an enum that could be shared across the entire package.

dwikler replied:
Great idea, more elegant, I implemented it.

sjswerdloff commented on tdwii_plus_examples/cli/upschangestatescu.py:
```diff
@@ -0,0 +1,128 @@
+        if args.state == "IN PROGRESS":
+            if args.transaction_uid:
+                parser.error("transaction_uid should not be provided when state is IN PROGRESS")
+            result = scu.claim_ups(ups_instance_uid)
```
how is the user of the CLI supposed to get the transaction UID?
It might be getting logged or printed somewhere underneath, but I think it would be useful for it to be printed out at this level for a successful claim_ups().

dwikler replied:
Definitely needed. I have added it.

sjswerdloff commented on tdwii_plus_examples/tdd/tdd_widget.py:
```diff
@@ -117,31 +117,35 @@ def _start_procedure(self):
             )
-        naction_scu.send_procedure_step_state_change(IN_PROGRESS, ups_uid=ups_uid)
-        self.current_transaction_uid = naction_scu.current_transaction_uid()
+        self.current_transaction_uid = naction_scu.claim_ups(sop_instance_uid=ups_uid)
```
This is what is "missing" from the CLI...

dwikler replied:
added.

sjswerdloff commented on tdwii_plus_examples/cli/upschangestatescu.py:
```diff
@@ -0,0 +1,128 @@
+
+
+if __name__ == "__main__":
+    main()
```
this always exits with exit code 0?
Shouldn't it exit with something non-zero if there is a failure or if the command line arguments aren't valid?

dwikler replied:
Added exit(1) in case of failure and adapted unit test to check the the right exit code value (0 for success, 1 for failure and 2 for parser error)

sjswerdloff commented on tdwii_plus_examples/tests/cli/test_upschangestatescu.py:
```diff
@@ -0,0 +1,67 @@
+        self.mock_scu.verify.return_value = mock.Mock(status_category="Failure", status_description="Failed")
+        exit_code = self.run_main_with_args(["upschangestatescu", "127.0.0.1", "11114", "1.2.3.4.5", "--echo"])
+        self.mock_scu.verify.assert_called_once()
+        self.assertEqual(exit_code, 0)
```
I'm a bit confused, because in run_main_with_args() you are providing for a non-zero exit code, but that doesn't seem like it will happen unless you have an uncaught exception.  If this is testing the library and you are looking at the return value then I think you need to assert that you got that return value.  If it's testing the CLI, I think you want the CLI to return a non-zero exit value if whatever the user was trying to do failed (including providing a bad command line).

dwikler replied:
solved. See previous comment.

sjswerdloff commented on tdwii_plus_examples/tests/test_cstorescu_unit.py:
```diff
@@ -30,7 +30,9 @@ def test_store_instances(self, mock_associate, mock_handle_response):
         """Test storing multiple DICOM instances."""
         # Create 2 empty instances
         ds_1 = Dataset()
+        ds_1.SOPClassUID = UID("1.2.840.10008.5.1.4.1.1.2")  # CT Image Storage SOP Class UID
```
I think you could either extract to constants, or use the constants provided in pynetdicom:
https://pydicom.github.io/pynetdicom/stable/reference/sop_classes.html#module-pynetdicom.sop_class

dwikler replied:
done.

sjswerdloff commented on tdwii_plus_examples/tests/test_upspullnactionscu_integration.py:
```diff
@@ -0,0 +1,171 @@
+        shutil.rmtree(cls.temp_dir)
+        cls.test_logger.info(f"Removed the temp directory: {cls.temp_dir} and its contents: {files_in_temp_dir}")
+
+        # Try to gracefully terminate the process.
```
wouldn't it be better to terminate the process first and then clean out the temporary directory and it's contents?
Unless the process (SCP?) is not using the temporary directory or it's contents.

dwikler replied:
Yes, modified.

sjswerdloff reviewed (changes requested):
I'd like to have a bit of discussion around some of my comments (even if you don't agree with any of them).

Also, With pytest you can have custom marks
e.g.:
@pytest.mark.integration
@pytest.mark.slow

So that you can separate out what gets tested on the command line.
One can also put integration tests in a separate directory from unit tests (and then choose which directories get listed for testing).  I'm not sure if unittest has the same kind of decorators available, but the separate directories should work for almost any test framework.

In the short term, I don't think it's a big deal.
In the longer term, if we want to test the SCUs against 3rd party SCPs (that might involve a fair amount of compute/time, e.g. a dockerized version of dcm4chee ) "prior to release" (but not on every PR), it might be beneficial.

dwikler commented:
I addressed all comments except the split between unit and integration tests. It seems unittest does not provide such decorators but there is a way to group tests in test suites. I agree that optimizing the CI test pipeline will be required at some point.

sjswerdloff commented on tdwii_plus_examples/tests/test_cstorescu_unit.py:
```diff
@@ -168,7 +174,7 @@ def test_set_contexts_valid(self):
     def test_set_contexts_invalid(self):
         """Test set_contexts with an invalid context."""
-        qr_abstract__syntax = "1.2.840.10008.5.1.4.1.2.2.1"  # not a storage context
+        qr_abstract__syntax = StudyRootQueryRetrieveInformationModelFind  # not a storage context
```
did you really intend to have __ (double underscore) in the variable name?

dwikler replied:
Definitely a typo, I just corrected it.

sjswerdloff reviewed (approved):
LGTM.
If you don't want to change the variable name with the double underscore, that's fine.
I think we should spend some time providing typing throughout (split it ~ 50/50 between us?) and use pyright or something like it to help us stay on top of that.
I'm very happy that you updated the _association() return values with boolean, NamedTuple.  It did however highlight the value of extensive typing (at least for code review).
